### ISSUE TITLE
 Cannot reuse already patched geometry.

### ISSUE THREAD

**Issue by OhBonsai** [agargaro/instanced-mesh#19]
Thank you for creating such a great library. While using it, I encountered an error when loading a glb file in r3f.
```
chunk-HM6WVW62.js?v=9216be66:19381 Uncaught Error: Cannot reuse already patched geometry.
    at new Gt (@three__ez_instanced-mesh.js?v=80ddd4ad:798:708)
    at InstancedEcd (InstancedEcd.jsx:10:25)
    at renderWithHooks (chunk-VNKMZVZQ.js?v=9216be66:5896:26)
    at mountIndeterminateComponent (chunk-VNKMZVZQ.js?v=9216be66:9379:21)
    at beginWork (chunk-VNKMZVZQ.js?v=9216be66:10362:22)
    at beginWork$1 (chunk-VNKMZVZQ.js?v=9216be66:14192:22)
    at performUnitOfWork (chunk-VNKMZVZQ.js?v=9216be66:13668:20)
    at workLoopSync (chunk-VNKMZVZQ.js?v=9216be66:13608:13)
    at renderRootSync (chunk-VNKMZVZQ.js?v=9216be66:13587:15)
    at recoverFromConcurrentError (chunk-VNKMZVZQ.js?v=9216be66:13177:28)
```

### My code
```
import { InstancedMesh2 } from '@three.ez/instanced-mesh';
import {useThree} from "@react-three/fiber";
import { useGLTF } from '@react-three/drei'
import * as THREE from 'three';

export default function InstancedEcd() {
  const { nodes, materials } = useGLTF('/draco/lingmou.glb')
  const renderer = useThree((state)=>state.gl)
  const material = new THREE.MeshBasicMaterial({color: 0xffffff});
  const instancedMesh = new InstancedMesh2(
    renderer, 10000, nodes.chasis.geometry,   new THREE.MeshNormalMaterial()
  )

  instancedMesh.updateInstances((obj, index) => {
    obj.position.z = index % 100;
    obj.position.x = index / 100;
    obj.rotateY(Math.PI);
  });
  instancedMesh.computeBVH();

  return <primitive object={instancedMesh} />


}
```

### **my enviroment**
```
macbook & chrome
    "@react-three/drei": "^9.99.7",
    "@react-three/fiber": "^8.15.19",
    "@three.ez/instanced-mesh": "^0.2.6",
    "three": "^0.162.0",

```

**Comment by agargaro** [comment]
Hello, thank you! 😄 

Geometries and materials cannot be shared with multiple `Mesh` / `InstancedMesh` because they are modified by adding unique data to manage the various instances.

In the case of geometries, an `InstancedBufferAttribute` is added to manage the indices that need to be rendered.

This can be solved by [cloning geometries](https://threejs.org/docs/#api/en/core/BufferGeometry.clone) and materials that are reused. 

```
const instancedMesh = new InstancedMesh2(
  renderer, 10000, nodes.chasis.geometry.clone(), new THREE.MeshNormalMaterial()
)
 ```

**Comment by agargaro** [comment]
@OhBonsai 
This needs to be handled better, I thank you for opening the bug. I will do a PR in the next few days :)

**Comment by agargaro** [comment]
From the next version, it will no longer be necessary to manually clone the geometry.

**Comment by jamespohalloran** [comment]
I'm running into this error as well. Started working from the latest version (0.2.10). I have some custom logic to collect / merge all my child meshes, maybe that doesn't work with this library?

Here's how Im loading my models:
```ts
    private _createInstancedMesh(platformMeta: { id: string; count: number; gltf: GLTF }) {
        if (platformMeta.gltf) {
            const threeModel: Group = platformMeta.gltf;

            const meshes = collectMeshes(threeModel);

            const geometries: BufferGeometry[] = meshes.map((mesh) => mesh.geometry);
            const mergedGeometry = mergeGeometries(geometries, true);

            if (threeModel.children && threeModel.children[0]) {
                // Create the InstancedMesh with the loaded model
                const platformMesh = new InstancedMesh2(
                    Game!.Renderer!.Renderer,
                    platformMeta.count,
                    mergedGeometry,
                    // if i just pass in a single material, this works.
                    // new MeshNormalMaterial()
                     meshes.map((mesh) => {
                         const material = mesh.material;
                         if (Array.isArray(material)) {
                             material.forEach((mat) => {
                               mat.side = FrontSide;
                             });
                         } else {
                             material.side = FrontSide;
                         }
                         return material;
                     })
                );
                
                return platformMesh;
            }
        } else {
            throw new Error(`Failed to load model for ${platformMeta.id}`);
        }
    }
  ```
  
P.s, I appreciate you making this library public! I've tested with just passing in a single material and I'm getting some great performance gains already, would just love to sort out using multiple materials on my merged geometry.

**Comment by jamespohalloran** [comment]
> I'm running into this error as well. Started working from the latest version (0.2.10). I have some custom logic to collect / merge all my child meshes, maybe that doesn't work with this library?

Sorry dug in deeper and I should expand:

I understand now that the issue here was several sub-meshes had the same material. When I filtered out unique materials (by name) I then got:
> SortingUtils.ts:18 Uncaught Error: Multi material is not supported.

HOWEVER! When I remove the custom sort I was also doing:
```
platformMesh.customSort = createRadixSort(platformMesh);
```
Things work! (Sorry, this should perhaps be its own ticket knowing the actual solution)

**Comment by agargaro** [comment]
> P.s, I appreciate you making this library public! I've tested with just passing in a single material and I'm getting some great performance gains

I'm glad 😄 


> platformMesh.customSort = createRadixSort(platformMesh);

This could be better supported if all materials are transparent or not. Is this your case?
If you'd like to help, would you like to create an example for me with radix sort + multimaterial?

**Comment by jamespohalloran** [comment]
> > P.s, I appreciate you making this library public! I've tested with just passing in a single material and I'm getting some great performance gains
> 
> I'm glad 😄
> 
> > platformMesh.customSort = createRadixSort(platformMesh);
> 
> This could be better supported if all materials are transparent or not. Is this your case? If you'd like to help, would you like to create an example for me with radix sort + multimaterial?
> 
> Anyway, it would be better to open a separate issue to discuss it :)

Will do! Will be tied up for the rest of the week but will try and carve out some time for this. For now I don't need radix sort. I don't believe any of the materials are transparent, but I will check on that as well

**Comment by agargaro** [comment]
I made a fix to use radix sort, but it will be available with version 0.3.0 :)
